torchmetrics/functional/text/chrf.py

```diff
@@ -0,0 +1,717 @@
+# Link:
+
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+# Copyright 2017 Maja Popovic
+
+# The program is distributed under the terms
+# of the GNU General Public Licence (GPL)
+
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+
+from collections import defaultdict
+from typing import Dict, List, Optional, Sequence, Tuple, Union
+
+import torch
+from torch import Tensor, tensor
+
+_EPS_SMOOTHING = tensor(1e-16)
+# Taken from https://github.com/mjpost/sacrebleu/blob/master/sacrebleu/metrics/chrf.py
+_PUNCTUATIONS = set("!\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~")
+
+
+def _prepare_n_grams_dicts(
+    n_char_order: int, n_word_order: int
+) -> Tuple[
+    Dict[int, Tensor], Dict[int, Tensor], Dict[int, Tensor], Dict[int, Tensor], Dict[int, Tensor], Dict[int, Tensor]
+]:
+    """Prepare dictionaries dictionaries with default zero values for total reference, hypothesis and matching
+    character and word n-grams.
+
+    Args:
+        n_char_order:
+            A character n-gram order.
+        n_word_order:
+            A word n-gram order.
+
+    Return:
+        Dictionaries with default zero values for total reference, hypothesis and matching character and word
+        n-grams.
+    """
+    total_ref_char_n_grams: Dict[int, Tensor] = {n + 1: tensor(0.0) for n in range(n_char_order)}
+    total_ref_word_n_grams: Dict[int, Tensor] = {n + 1: tensor(0.0) for n in range(n_word_order)}
+    total_hyp_char_n_grams: Dict[int, Tensor] = {n + 1: tensor(0.0) for n in range(n_char_order)}
+    total_hyp_word_n_grams: Dict[int, Tensor] = {n + 1: tensor(0.0) for n in range(n_word_order)}
+    total_matching_char_n_grams: Dict[int, Tensor] = {n + 1: tensor(0.0) for n in range(n_char_order)}
+    total_matching_word_n_grams: Dict[int, Tensor] = {n + 1: tensor(0.0) for n in range(n_word_order)}
+
+    return (
+        total_ref_char_n_grams,
+        total_ref_word_n_grams,
+        total_hyp_char_n_grams,
+        total_hyp_word_n_grams,
+        total_matching_char_n_grams,
+        total_matching_word_n_grams,
+    )
+
+
+def _defaultdict_of_tensors_tuple_keys() -> Dict[Tuple[str, ...], Tensor]:
+    """A wrapper for creating `defaultdict` with key type of `Tuple[str, ...]` and initialized with a zero
+    tensor."""
+
+    def zero_tensor() -> Tensor:
+        return tensor(0.0)
+
+    return defaultdict(zero_tensor)
+
+
+def _defaultdict_of_tensors_int_keys() -> Dict[int, Tensor]:
+    """A wrapper for creating `defaultdict` with key type of `int` and initialized with a zero tensor."""
+
+    def zero_tensor() -> Tensor:
+        return tensor(0.0)
+
+    return defaultdict(zero_tensor)
```

Comment:
why can't we use `defaultdict(lambda : tensor(0.0))` for that? This function is only meant for initialization right? Is it ever pickled?

Reply:
Yes, that's a definitely good idea to use lambda functions instead! :]